src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+function_pointer_restrictionst::restrictionst
+parse_function_pointer_restrictions_from_command_line(
+  const std::list<std::string> &restriction_opts)
+{
+  auto function_pointer_restrictions =
+    function_pointer_restrictionst::restrictionst{};
+  for(auto const &restriction_opt : restriction_opts)
+  {
+    // the format for restrictions is <pointer_name>/<target[,more_targets]*>
+    // exactly one pointer and at least one target
+    auto const pointer_name_end = restriction_opt.find('/');
+    auto const restriction_format_message =
+      "the format for restrictions is "
+      "<pointer_name>/<target[,more_targets]*>";
+    if(pointer_name_end == std::string::npos)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "couldn't find '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    if(pointer_name_end == restriction_opt.size())
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "couldn't find names of targets after '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    if(pointer_name_end == 0)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "couldn't find target name before '/' in `" + restriction_opt + "'",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
+    }
+    auto const pointer_name = restriction_opt.substr(0, pointer_name_end);
+    auto const target_names_substring =
+      restriction_opt.substr(pointer_name_end + 1);
+    auto const target_name_strings = split_string(target_names_substring, ',');
+    if(target_name_strings.size() == 1 && target_name_strings[0].empty())
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "missing target list for function pointer restriction " + pointer_name,
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    auto const target_names = ([&target_name_strings] {
+      auto target_names = std::unordered_set<irep_idt>{};
+      std::transform(
+        target_name_strings.begin(),
+        target_name_strings.end(),
+        std::inserter(target_names, target_names.end()),
+        string2id);
+      return target_names;
+    })();
+    for(auto const &target_name : target_names)
+    {
+      if(target_name == ID_empty_string)
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "leading or trailing comma in restrictions for `" + pointer_name +
+            "'",
+          "--" RESTRICT_FUNCTION_POINTER_OPT,
+          restriction_format_message);
+      }
+    }
+    if(!function_pointer_restrictions
+          .emplace(irep_idt{pointer_name}, target_names)
+          .second)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "function pointer restriction for `" + pointer_name +
+          "' was specified twice",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
```

Comment:
:no_entry_sign: Pull out into a function that takes a string and returns a data structure containing pointer_name, target_names and unit test to cover all the awkward cases here (before swapping it to use regex if you do choose to do that)

Reply:
Done